fgpyo/sam/__init__.py

```diff
@@ -584,3 +586,110 @@ def set_pair_info(r1: AlignedSegment, r2: AlignedSegment, proper_pair: bool = Tr
     insert_size = isize(r1, r2)
     r1.template_length = insert_size
     r2.template_length = -insert_size
+
+
+@attr.s(auto_attribs=True, frozen=True)
+class Template:
+    """A container for alignment records corresponding to a single sequenced template
+    or insert.
+
+    Attributes:
+        name: the name of the template/query
+        r1: Primary alignment for read 1, or None if there is none
+        r2: Primary alignment for read 2, or None if there is none
+        r1_supplementals: Supplementary alignments for read 1
+        r2_supplementals: Supplementary alignments for read 2
+        r1_secondaries: Secondary (non-primary) alignments for read 1
+        r2_secondaries: Secondary (non-primary) alignments for read 2
+    """
+
+    name: str
+    r1: Optional[AlignedSegment] = None
+    r2: Optional[AlignedSegment] = None
+    r1_supplementals: List[AlignedSegment] = []
+    r2_supplementals: List[AlignedSegment] = []
+    r1_secondaries: List[AlignedSegment] = []
+    r2_secondaries: List[AlignedSegment] = []
+
+    @staticmethod
+    def iterator(alns: Iterator[AlignedSegment]) -> Iterator["Template"]:
+        """Returns an iterator over templates. Assumes the input iterable is queryname grouped,
+        and gathers contiguous runs of records sharing a common query name into templates."""
+        return TemplateIterator(alns)
+
+    @staticmethod
+    def build(recs: Iterable[AlignedSegment]) -> "Template":
+        """Build a template from a set of records all with the same queryname."""
+        r1 = None
+        r2 = None
+        r1_supplementals: List[AlignedSegment] = []
+        r2_supplementals: List[AlignedSegment] = []
+        r1_secondaries: List[AlignedSegment] = []
+        r2_secondaries: List[AlignedSegment] = []
+
+        for rec in recs:
+            is_r1 = not rec.is_paired or rec.is_read1
+
+            if not rec.is_supplementary and not rec.is_secondary:
+                if is_r1:
+                    assert r1 is None, f"Multiple R1 primary reads found in {recs}"
+                    r1 = rec
+                else:
+                    assert r2 is None, f"Multiple R2 primary reads found in {recs}"
+                    r2 = rec
+            elif rec.is_supplementary:
+                if is_r1:
+                    r1_supplementals.append(rec)
+                else:
+                    r2_supplementals.append(rec)
+            if rec.is_secondary:
+                if is_r1:
+                    r1_secondaries.append(rec)
+                else:
+                    r2_secondaries.append(rec)
+
+        return Template(
+            name=next(iter(recs)).query_name,
```

Comment:
Wait, how will this work if `recs` is an iterator?  You've exhausted `recs` in the for loop.  Also consider a better error message if `recs` is empty.

Reply:
Could you add a test for `build` where an `iter` is passed?

Reply:
I've changed it.  But f' python.  `Iterator` should not implement `Iterable`.  Sigh.